pkg/util/portforward.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to create listener: Error %s", err)
+	}
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				if !strings.Contains(strings.ToLower(err.Error()), "use of closed network connection") {
+					runtime.HandleError(fmt.Errorf("error accepting connection on port %d: %v", r.targetPort, err))
+				}
+			}
+			go func() {
+				if err := r.handle(conn); err != nil {
+					runtime.HandleError(fmt.Errorf("error handling connection: %v", err))
+				}
+			}()
+		}
+	}()
+	return func() { listener.Close() }, nil
+
+}
+
+func (r *roundRobin) handle(conn net.Conn) error {
+	transport, upgrader, err := spdy.RoundTripperFor(r.restConfig)
+	if err != nil {
+		return err
+	}
+	nativeClient, err := kubernetes.NewForConfig(r.restConfig)
+	if err != nil {
+		return err
+	}
+	podList, err := nativeClient.CoreV1().
+		Pods(r.proxyServerNamespace).
+		List(context.TODO(), metav1.ListOptions{
+			LabelSelector: r.podSelector,
+		})
+	if err != nil {
+		return err
+	}
+
+	podIdx := rand.Intn(len(podList.Items))
+	pod := podList.Items[podIdx]
+	req := nativeClient.RESTClient().
+		Post().
+		Prefix("api", "v1").
+		Resource("pods").
+		Namespace(r.proxyServerNamespace).
+		Name(pod.Name).
+		SubResource("portforward")
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	streamConn, _, err := dialer.Dial(PortForwardProtocolV1Name)
+	if err != nil {
+		return err
+	}
+	defer streamConn.Close()
+
+	// create error stream
+	headers := http.Header{}
+	headers.Set(v1.StreamType, v1.StreamTypeError)
+	headers.Set(v1.PortHeader, fmt.Sprintf("%d", r.targetPort))
+	headers.Set(v1.PortForwardRequestIDHeader, strconv.Itoa(r.reqId))
+	r.reqId++
+	errorStream, err := streamConn.CreateStream(headers)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("error creating error stream for port %d -> %d: %v", r.targetPort, r.targetPort, err))
+		return err
+	}
+	// we're not writing to this stream
+	errorStream.Close()
+
+	errorChan := make(chan error)
+	go func() {
+		message, err := ioutil.ReadAll(errorStream)
+		switch {
+		case err != nil:
+			errorChan <- fmt.Errorf("error reading from error stream for port %d -> %d: %v", r.targetPort, r.targetPort, err)
+		case len(message) > 0:
+			errorChan <- fmt.Errorf("an error occurred forwarding %d -> %d: %v", r.targetPort, r.targetPort, string(message))
+		}
+		close(errorChan)
+	}()
+
+	// create data stream
+	headers.Set(v1.StreamType, v1.StreamTypeData)
+	dataStream, err := streamConn.CreateStream(headers)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("error creating forwarding stream for port %d -> %d: %v", r.targetPort, r.targetPort, err))
+		return err
+	}
+
+	localError := make(chan struct{})
+	remoteDone := make(chan struct{})
+
+	go func() {
+		// Copy from the remote side to the local port.
+		if _, err := io.Copy(conn, dataStream); err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
```

Comment:
why do we need to exclude the closed error?

Reply:
https://github.com/kubernetes/kubernetes/blob/cd6ffff85d257ff9067d59339f2ffdbcc66dc164/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L309-L310

actually this line is copied from the original client-go's port-forward server. for now the client-go library cannot be reused by cluster-proxy due to some limitation